Add documentation for Clone Snapshot Java API (#70720)
elastic/elasticsearch: docs/java-rest/high-level/snapshot/clone_snapshot.asciidoc

[[java-rest-high-snapshot-clone-snapshot]]
=== Clone Snapshot API

> The Clone Snapshot API allows to clone part or all of a snapshot into a new snapshot.

Reviewer: ```suggestion
The Clone Snapshot API  clones part or all of a snapshot into a new snapshot.
```

[[java-rest-high-snapshot-clone-snapshot-request]]

> ==== Clone Snapshot Request

Reviewer: nit
```suggestion
==== Request
```

A `CloneSnapshotRequest`:

["source","java",subs="attributes,callouts,macros"]
--------------------------------------------------
include-tagged::{doc-tests}/SnapshotClientDocumentationIT.java[clone-snapshot-request]
--------------------------------------------------

==== Indices to Clone

The `indices` property is used to provide the list of

> indices that should be included in the snapshot:

Reviewer 2: provide the list of indices from the source snapshot to be included in the snapshot clone

> The `indices` property is used to provide the list of
> indices that should be included in the snapshot:

Reviewer: ```suggestion
Use `indices` to specify a list of indices from the source snapshot to include
in the snapshot clone:
```

["source","java",subs="attributes,callouts,macros"]
--------------------------------------------------
include-tagged::{doc-tests}/SnapshotClientDocumentationIT.java[clone-snapshot-request-indices]
--------------------------------------------------

> <1> Request that Elasticsearch only includes "test_index" from the source snapshot.

Reviewer: ```suggestion
<1> Include only `test_index` from the source snapshot.
```

==== Index Settings and Options

You can also customize index settings and options when cloning a snapshot:

["source","java",subs="attributes,callouts,macros"]
--------------------------------------------------
include-tagged::{doc-tests}/SnapshotClientDocumentationIT.java[clone-snapshot-request-index-settings]
--------------------------------------------------
<1> Set `IndicesOptions.Option.IGNORE_UNAVAILABLE` in `#indicesOptions()` to
    have the clone succeed even if indices are missing in the source snapshot.

==== Further Arguments

> The following arguments can optionally be provided:

Reviewer: ```suggestion
You can also provide the following optional arguments:
```

passed to the asynchronous method:

["source","java",subs="attributes,callouts,macros"]
--------------------------------------------------
include-tagged::{doc-tests}/SnapshotClientDocumentationIT.java[clone-snapshot-execute-async]
--------------------------------------------------
<1> The `CloneSnapshotRequest` to execute and the `ActionListener`
to use when the execution completes

The asynchronous method does not block and returns immediately. Once it is
completed the `ActionListener` is called back using the `onResponse` method
if the execution successfully completed or using the `onFailure` method if
it failed.

A typical listener for `AcknowledgedResponse` looks like:

["source","java",subs="attributes,callouts,macros"]
--------------------------------------------------
include-tagged::{doc-tests}/SnapshotClientDocumentationIT.java[clone-snapshot-execute-listener]
--------------------------------------------------
<1> Called when the execution is successfully completed. The response is
    provided as an argument.
<2> Called in case of a failure. The raised exception is provided as an argument.

[[java-rest-high-cluster-clone-snapshot-response]]

> ==== Clone Snapshot Response

Reviewer: nit
```suggestion
==== Response
```

> The returned `AcknowledgedResponse` contains a value on whether or not the request
> was received:

Reviewer: ```suggestion
`AcknowledgedResponse` indicates whether the request was received:
```

["source","java",subs="attributes,callouts,macros"]
--------------------------------------------------
include-tagged::{doc-tests}/SnapshotClientDocumentationIT.java[clone-snapshot-response]
--------------------------------------------------

> <1> A boolean value of `true` if the clone successfully completed, `false` otherwise.

Reviewer: Nit
```suggestion
<1> A boolean value of `true` if the clone successfully completed. Otherwise, the value is `false`.
```
